### PR TITLE
Use ._tshift internally for datetimelike ops

### PR DESCRIPTION
**Pull request by TomAugspurger**
In preperation for PeriodArray / DatetimeArray / TimedeltaArray.

Index.shift has a different meaning from ExtensionArray.shift.

- Index.shift pointwise shifts each element by some amount
- ExtensionArray.shift shits the *position* of each value in the array
  padding the end with NA

This is going to get confusing. This PR tries to avoid some of that by
internally using a new `_tshift` method (time-shift) when we want to do pointwise
shifting of each value. Places that know they want that behavior (like in the
datetimelike ops) should use that.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -553,6 +553,23 @@ def shift(self, periods, freq=None):
         --------
         Index.shift : Shift values of Index.
         """
+        return self._tshift(periods=periods, freq=freq)
+
+    def _tshift(self, periods, freq=None):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: why don't we call this ._time_shift or ._period_shift

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -553,6 +553,23 @@ def shift(self, periods, freq=None):
...
+        """
+        Shift each value by `periods`.
+
+        Note this is different from ExtensionArray.shift, which
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: how is this relevant here? this is simply a re-definition of what shift means. I understand the desire to make things clear, but this has *always* been the case, for example in comparing a DTI to an Index.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: if you are actually going to change this, then this should be much more consistent across the code base, e.g. does internals need changing, what about the Index classes?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: we have a note to deprecate tshift entirely. I don't think this is a good name at all.

**Comment by TomAugspurger** [comment]
Renamed to `_time_shift`.

> if you are actually going to change this, then this should be much more consistent across the code base, e.g. does internals need changing, what about the Index classes?

Do you have particular places in mind?

**Comment by TomAugspurger** [comment]
All green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**